### PR TITLE
Add endpoint-specific gnix_fab_req freelist (and new criterion test for endpoints)

### LINKED ISSUE #100: need simple freelist for gnix_fab_reqs
We need a simple freelist (probably associated with the `gnix_domain_t` struct) to manage `gnix_fab_reqs`.   An initial set of may 100 elements could be added to the freelist, with additional
elements being allocated out of the heap.  This initial set size should ultimately be controllable using an `ops_open` method on the domain handle. 

This task needs to be completed as part of the solution for issue #5.

### PR DESCRIPTION
**Pull request by sungeunchoi**
I also cleaned up the error exiting in gnix_ep_open().

Note: There is an issue with allocating more than 9 endpoints at a time that needs to be investigated.

@hppritcha @ztiffany @jswaro 

Closes #100 

Signed-off-by: Sung-Eun Choi sungeunchoi@users.noreply.github.com

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/63/
Test PASSed.

**Comment by jswaro** [comment]
RE: endpoint allocation

Does it refuse to allocate the endpoint? What is the error code?

**Comment by sungeunchoi** [comment]
`GNI_CqCreate` fails with GNI_RC_ERROR_RESOURCE for the rx_cq_blk (which is the 4th call to `GNI_CqCreate` in `gnix_nic_alloc`.

**Comment by hppritcha** [comment]
this looks like it needs to be rebased.

File: prov/gni/include/gnix.h [hunk 1]
```diff
@@ -265,8 +266,9 @@ struct gnix_fid_ep {
 	int (*rx_progress_fn)(struct gnix_fid_ep *, gni_return_t *rc);
 	int enabled;
 	int no_want_cqes;
-	/* num. active fab_reqs associated with this ep */
+	/* num. active put (write) fab_reqs associated with this ep */
```
Comment (hppritcha) [inline prov/gni/include/gnix.h]: The active count also tracks get (read) reqs.

**Comment by sungeunchoi** [comment]
@hppritcha 
I rebased, and also cleaned up the error return path in gnix_ep_open().  FYI, there was a bad check on line 730 (must have missed it during my review of #163).

**Comment by hppritcha** [comment]
:100: 

I'm good with this assuming jenkins passes it.

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/69/
Test PASSed.

**Pull request merged**